### PR TITLE
GH-37212: [C++] IO: BufferReader always owned buffer

### LINKED ISSUE #37212: [C++] BufferReader should zero-copy if and only if given a Buffer, not a pointer and length
### Describe the enhancement requested

BufferReader can be constructed from a pointer and length:

https://github.com/apache/arrow/blob/0e677d27ae8ed489f2fa6f5e95b0e498319a887e/cpp/src/arrow/io/memory.cc#L277-L279

In that case, it will happily hand you a Buffer backed by that pointer:

https://github.com/apache/arrow/blob/0e677d27ae8ed489f2fa6f5e95b0e498319a887e/cpp/src/arrow/io/memory.cc#L353

However, I don't think the APIs make it clear that this ties the lifetime of any data read from the reader to the lifetime of the input. It would be clearer if it would copy in this case, and require you to explicitly construct BufferReader from a Buffer if you want to enable zero-copy.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Previously, when input an non-owned string, `arrow::io::BufferReader` would zero-copy it. It would cause lifetime problem. This patch force it copy in this case.

### What changes are included in this PR?

1. Force `arrow::io::BufferReader` to be "owned"
2. Add a ctor for `BufferReader(std::string)`, and remove non-owned ctors

### Are these changes tested?

Yes

### Are there any user-facing changes?

User can use BufferReader safely, but might get little more memcopy

* Closes: #37212

### CODE REVIEW

**Comment by mapleFU** [comment]
Seems I meet a problem like https://github.com/nlohmann/json/issues/586

Should I use `BufferReader::From(std::string)` like `::arrow::Buffer`, or using the `json` like method to solve this?

**Review by bkietz — COMMENTED** [review]
This approach will work, but I'm reluctant to introduce a code path where we allocate-and-copy on read in a class from which users could otherwise reasonably expect trivial IO costs.

As a potential alternative:

These non-owning constructors don't seem widely used; mostly tests and deserialization of FunctionOptions. Perhaps instead demoting BufferReader to only sometimes zero-copy, it'd be better to delete these constructors and require users to own the buffers to be read (probably doing a single copy all at once on construction instead of one copy per read).

@lidavidm

**Comment by lidavidm** [comment]
Also reasonable to me, though you'd have to go through a deprecation cycle I believe.

**Comment by lidavidm** [comment]
Or you could have those constructors do the copy for you (and if you really want to zero-copy from a temporary, you can always explicitly construct the Buffer yourself)

**Comment by mapleFU** [comment]
> This approach will work, but I'm reluctant to introduce a code path where we allocate-and-copy on read in a class from which users could otherwise reasonably expect trivial IO costs.

I also vote for this, let me implement it

File: cpp/src/arrow/compute/function_internal.cc [hunk 1]
```diff
@@ -83,7 +83,8 @@ Result<std::unique_ptr<FunctionOptions>> GenericOptionsType::Deserialize(
 
 Result<std::unique_ptr<FunctionOptions>> DeserializeFunctionOptions(
     const Buffer& buffer) {
-  io::BufferReader stream(buffer);
+  // FIXME: Would ToString to heavy?
```
Comment (mapleFU) [inline cpp/src/arrow/compute/function_internal.cc:86]: I wonder would this be a bit heavy? Should I implement a `DeserializeFunctionOptions(const std::shared_ptr<Buffer>& buffer)`

**Comment by mapleFU** [comment]
@bkietz @lidavidm I've remove the non-owned style, and find https://github.com/apache/arrow/pull/37271#discussion_r1300855362 might introduce cost, would you mind take a look?

File: cpp/src/arrow/compute/function_internal.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/function_internal.cc:86]: You can explicitly `std::make_shared<Buffer>(data_, nbytes)` to preserve the prior behavior.

Comment (mapleFU) [reply cpp/src/arrow/compute/function_internal.cc:86]: sorry I'm not fully understand. 

```c++
io::BufferReader stream(buffer.ToString());
```

is same as `io::BufferReader`. I mean should change `DeserializeFunctionOptions(const Buffer& buffer)`  to `DeserializeFunctionOptions(const std::shared_ptr<Buffer>& buffer)`

Comment (lidavidm) [reply cpp/src/arrow/compute/function_internal.cc:86]: What I mean is that you can keep `const Buffer&` and just explicitly wrap it in a non-owning `shared_ptr<Buffer>` to pass to `BufferReader`, which keeps the old behavior (zero-copy without owning)

Comment (mapleFU) [reply cpp/src/arrow/compute/function_internal.cc:86]: I got it, thanks!

**Comment by mapleFU** [comment]
```
C:\projects\arrow>mamba update -q -y -c conda-forge --all   || exit /B 
Traceback (most recent call last):
  File "C:\Miniconda38-x64\Scripts\mamba-script.py", line 6, in <module>
    from mamba.mamba import main
  File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 49, in <module>
    import libmambapy as api
  File "C:\Miniconda38-x64\lib\site-packages\libmambapy\__init__.py", line 7, in <module>
    raise e
  File "C:\Miniconda38-x64\lib\site-packages\libmambapy\__init__.py", line 4, in <module>
    from libmambapy.bindings import *  # noqa: F401,F403
ImportError: DLL load failed while importing bindings: The specified module could not be found.
C:\projects\arrow>set lastexitcode=1 
C:\projects\arrow>set  1>C:\Users\appveyor\AppData\Local\Temp\1\tmp82BE.tmp 
C:\projects\arrow>echo C:\projects\arrow  1>C:\Users\appveyor\AppData\Local\Temp\1\tmp82BF.tmp 
C:\projects\arrow>exit /b 1 

```

No idea why it failed...

File: cpp/src/arrow/io/memory.h [hunk 2]
```diff
@@ -146,12 +146,9 @@ class ARROW_EXPORT BufferReader
     : public internal::RandomAccessFileConcurrencyWrapper<BufferReader> {
  public:
   explicit BufferReader(std::shared_ptr<Buffer> buffer);
-  explicit BufferReader(const Buffer& buffer);
-  BufferReader(const uint8_t* data, int64_t size);
 
-  /// \brief Instantiate from std::string or std::string_view. Does not
-  /// own data
-  explicit BufferReader(std::string_view data);
```
Comment (lidavidm) [inline cpp/src/arrow/io/memory.h:154]: Maybe we should deprecate instead of remove?

Comment (lidavidm) [reply cpp/src/arrow/io/memory.h:154]: We could even introduce static factories that have the same behavior as these constructors, but whose names explicitly indicate that they may be unsafe

Comment (mapleFU) [reply cpp/src/arrow/io/memory.h:154]: OK, seems that I should mark them deprecated, but keep them as before (including zero-copy)

Then how do I fix https://github.com/apache/arrow/pull/37271#issuecomment-1686585639 ? Using a `BufferReader::FromString`? Or other way?

Comment (lidavidm) [reply cpp/src/arrow/io/memory.h:154]: Ah, that's annoying.

`FromString` is probably best.

**Pull request closed without merging**